Extended side effect error messages

quantimnot opened the pull request:
Fixes #4210
Fixes #10712

Is this the right approach?
Any desired changes to make the messages clearer?
All feedback is welcome.

quantimnot committed:
Extended side effect error messages

timotheecour commented on compiler/sempass2.nim:
```diff
@@ -225,7 +225,17 @@ proc markGcUnsafe(a: PEffects; reason: PNode) =
+    if not a.inEnforcedNoSideEffects:
+      a.hasSideEffect = true
+      if a.owner.kind in routineKinds:
+        when reason is PNode:
+          if reason.kind == nkSym:
+            a.owner.sideEffectReasons.add reason.sym
+          else:
+            a.owner.sideEffectReasons.add newSym(skUnknown, a.owner.name, nextSymId a.c.idgen,
+                                              a.owner, reason.info, {})
+        else:
+          a.owner.sideEffectReasons.add reason
```
factor. eg:
var sym: PSym
when: ... else: ...
a.owner.sideEffectReasons.add sym

timotheecour commented on compiler/sempass2.nim:
```diff
@@ -1357,7 +1396,10 @@ proc trackProc*(c: PContext; s: PSym, body: PNode) =
     else:
-      localError(g.config, s.info, ("'$1' can have side effects" % s.name.s) & (g.config $ mutationInfo))
+      if hasMutationSideEffect:
+        localError(g.config, s.info, ("'$1' can have side effects" % s.name.s) & (g.config $ mutationInfo))
```
```nim
localError(g.config, s.info, "'$1' can have side effects$2" % [s.name.s, g.config $ mutationInfo])
```

ditto for other instances (eg in listSideEffects)

timotheecour commented on compiler/sempass2.nim:
```diff
@@ -259,6 +269,33 @@ proc listGcUnsafety(s: PSym; onlyWarning: bool; conf: ConfigRef) =
   listGcUnsafety(s, onlyWarning, cycleCheck, conf)
 
+proc listSideEffects(s: PSym; onlyWarning: bool; cycleCheck: var IntSet; conf: ConfigRef) =
+  for u in s.sideEffectReasons:
```
seems expensive; is that called in a compiles() context? if so, it needs to be disabled in the same way as extended diagnostics are skipped for type mismatch during `compiles()`

timotheecour commented on compiler/sempass2.nim:
```diff
@@ -259,6 +269,33 @@ proc listGcUnsafety(s: PSym; onlyWarning: bool; conf: ConfigRef) =
+          "'$#' has side effects as it performs an indirect call via '$#'" %
+          [s.name.s, u.name.s])
+      else:
+        message(conf, u.info, msgKind,
```
generate a single message instead and get rid of `onlyWarning: bool; ` which simplifies and renders better:
```nim
proc listSideEffects(result: var string, s: PSym; cycleCheck: var IntSet; conf: ConfigRef) = ... # ok, still recursive

var msg = ""
listSideEffects(msg, ...)
message(conf, n.info, errGenerated, msg)
```

```nim
proc baz(c: int) =
  echo c
proc bar(b: int) =
  baz(b)
func fn1(a: int) =
  bar(a)
proc main()=
  fn1(3)
main()
```
current:
```
tests/nim/all/t12516.nim(9, 1) template/generic instantiation from here
tests/nim/all/t12516.nim(5, 6) Warning: 'baz' has side effects as it calls 'echo' [SideEffect]
  proc baz(c: int) =
       ^
tests/nim/all/t12516.nim(9, 1) template/generic instantiation from here
tests/nim/all/t12516.nim(7, 6) Warning: 'bar' has side effects as it calls 'baz' [SideEffect]
  proc bar(b: int) =
       ^
tests/nim/all/t12516.nim(9, 6) Error: 'fn1' has side effects as it calls 'bar'
```

instead something like this is better:
```
tests/nim/all/t12516.nim(9, 1) template/generic instantiation from here
tests/nim/all/t12516.nim(9, 6) Error: 'fn1' has side effects as it calls 'bar': func fn1(a: int) =
  tests/nim/all/t12516.nim(7, 6) 'bar' has side effects as it calls 'baz': proc bar(b: int) =
  tests/nim/all/t12516.nim(5, 6) 'baz' has side effects as it calls 'echo':  proc baz(c: int) =
```

quantimnot replied:
I changed `listSideEffects` to compile a single message.
I'm unsure of the actual rendered messages though.
The rendering form that you state above shows all locations as errors when I use the nim vscode plugin. But I think it would be clearer if the other locations were hints.

What about something like:
```
<lineinfo> Error: `symbol` has side effects
  <lineinfo> Hint: `symbol` accesses/mutates/calls `reason`
    <lineinfo> Hint: `reason` declared here
```

timotheecour replied:
one problem is the other locations are not really hints, they're mascarading as such; hints go through a specific code path in the compiler, which generates color, a hint name etc.

not a big deal though, we can always improve this later

quantimnot replied:
Yeah, realized that it's not the same, but I wanted the Nim VSCode plugin to be able to parse it and then markup my code in the editor. VSCode has diagnostic message types of hint, warning and error. The Nim VSCode plugin can be changed to map something else to internal VSCode hints.

quantimnot committed:
Applied feedback:
- refactored `markSideEffect`
- refactored string interpolations
- single message
- skip diagnostics in `system.compiles` context

Other:
- started a test of diagnostic messages

[ci skip] Tests aren't updated yet because messaging isn't nailed down.

quantimnot commented:
I just found nim-lang/RFCs#323 and nim-lang/RFCs#324.

I'm not sure what to apply from those discussions.

quantimnot committed:
- Added hints of where for side effect call locations.
- Tried to clarify the reasons.

quantimnot commented:
tests/effects/tdiagnostic_messages.nim:
```nim
var globalVar = 0
proc indirectCallViaVarParam(call: var proc(): int {.nimcall.}): int =
  call()
proc indirectCallViaPointer(call: pointer): int =
  cast[ptr proc(): int](call)[]()
proc callWithSideEffects(): int =
  var p = proc (): int {.nimcall.} = 0
  discard indirectCallViaVarParam(p)
  discard indirectCallViaPointer(addr p)
  echo ""
  globalVar

func a: int =
  discard callWithSideEffects()
```

Here is the current state of rendering:

Terminal:
```
tdiagnostic_messages.nim(31, 6) Error: 'a' has side effects
  tdiagnostic_messages.nim(32, 30) Hint: 'a' calls 'callWithSideEffects' which has side effects
    tdiagnostic_messages.nim(24, 6) Hint: 'callWithSideEffects' called by a `noSideEffect` routine, but has side effects
      tdiagnostic_messages.nim(26, 34) Hint: 'callWithSideEffects' calls 'indirectCallViaVarParam' which has side effects
        tdiagnostic_messages.nim(20, 6) Hint: 'indirectCallViaVarParam' called by a `noSideEffect` routine, but has side effects
          tdiagnostic_messages.nim(21, 7) Hint: 'indirectCallViaVarParam' calls a routine with side effects via hidden pointer indirection
      tdiagnostic_messages.nim(27, 33) Hint: 'callWithSideEffects' calls 'indirectCallViaPointer' which has side effects
        tdiagnostic_messages.nim(22, 6) Hint: 'indirectCallViaPointer' called by a `noSideEffect` routine, but has side effects
          tdiagnostic_messages.nim(23, 32) Hint: 'indirectCallViaPointer' calls a routine with side effects via pointer indirection
      tdiagnostic_messages.nim(28, 8) Hint: 'callWithSideEffects' calls 'echo' which has side effects
        ../../lib/system.nim(2004, 6) Hint: 'echo' called by a `noSideEffect` routine, but has side effects
      tdiagnostic_messages.nim(29, 3) Hint: 'callWithSideEffects' accesses global state 'globalVar'
        tdiagnostic_messages.nim(19, 5) Hint: 'globalVar' accessed by a `noSideEffect` routine
```

VSCode:

![vscode](https://user-images.githubusercontent.com/54247259/124391609-65388800-dcbf-11eb-969f-43a3562e8fbb.png)

quantimnot committed:
fix tests

timotheecour commented on tests/effects/tdiagnostic_messages.nim:
```diff
@@ -0,0 +1,32 @@
+discard """
+  errormsg: "\'a\' can have side effects"
```
use
```
nimoutFull: true
nimout: '''
blah blah
```
same as in https://github.com/nim-lang/Nim/pull/18384
(and remove line, file, errormsg)

rationale:
* easier to copy paste when things change (a single block)
* ensures we collect all the ouptu

timotheecour commented on compiler/ast.nim:
```diff
@@ -857,6 +857,7 @@ type
     of routineKinds:
       #procInstCache*: seq[PInstantiation]
       gcUnsafetyReason*: PSym  # for better error messages wrt gcsafe
+      sideEffectReasons*: seq[(TLineInfo, PSym)]  # for better side effect error messages
```
instead of changing AST for something that's only used for error msgs, add a Table/HashSet/other to `TContext` (similar to unusedImports or importModuleMap etc, indexed via `PSym.id`)

timotheecour commented on compiler/sempass2.nim:
```diff
@@ -224,8 +224,25 @@ proc markGcUnsafe(a: PEffects; reason: PNode) =
 when true:
-  template markSideEffect(a: PEffects; reason: typed) =
-    if not a.inEnforcedNoSideEffects: a.hasSideEffect = true
+  template markSideEffect(a: PEffects; reason: typed; useLoc: TLineInfo) =
```
turn this into a proc/generic, something like this:
```nim
proc markSideEffect[T](a: PEffects; reason: T; useLoc: TLineInfo) = ...
or even:
proc markSideEffect(a: PEffects; reason: PNode | something; useLoc: TLineInfo) = ...
```

timotheecour commented on compiler/sempass2.nim:
```diff
@@ -224,8 +224,25 @@ proc markGcUnsafe(a: PEffects; reason: PNode) =
+          sym = reason
+        a.owner.sideEffectReasons.add (useLoc, sym)
 else:
   template markSideEffect(a: PEffects; reason: typed) =
```
remove this disabled template, and add this at the right place:
```nim
when false: markGcUnsafe(a, reason)
```
which preserves the orginial intent without introducing code duplication

quantimnot replied:
What is the intent of `when false`?
Keep code around that maybe needed later on?

timotheecour replied:
yes, and ensures it's still highlighted and parseable, unlike comments

timotheecour commented on compiler/sempass2.nim:
```diff
@@ -224,8 +224,25 @@ proc markGcUnsafe(a: PEffects; reason: PNode) =
+          if reason.kind == nkSym:
+            sym = reason.sym
+          else:
+            var kind = skUnknown
+            case reason.kind
+            of nkHiddenDeref:
+              kind = skParam
+            else: discard
```
```suggestion
            let kind = if reason.kind == nkHiddenDeref: skParam else: skUnknown
```

timotheecour commented on compiler/sempass2.nim:
```diff
@@ -259,6 +276,29 @@ proc listGcUnsafety(s: PSym; onlyWarning: bool; conf: ConfigRef) =
+    if u != nil and not cycleCheck.containsOrIncl(u.id):
+      case u.kind
+      of skLet, skVar:
+        add("$# Hint: '$#' accesses global state '$#'\n" % [conf $ useLineInfo, s.name.s, u.name.s])
```
use `addf` which appends in place (can be done even if `template add(msg; level = indentLevel)` is used under the scene)
also, name `template add(msg; level = indentLevel)` differently to improve readability and avoid confusion

timotheecour commented on compiler/sempass2.nim:
```diff
@@ -259,6 +276,29 @@ proc listGcUnsafety(s: PSym; onlyWarning: bool; conf: ConfigRef) =
 
+proc listSideEffects(result: var string; s: PSym; cycleCheck: var IntSet; conf: ConfigRef; indentLevel: int) =
+  template add(msg; level = indentLevel) =
+    result.add repeat("  ", level) & msg
```
```nim
    result.add repeat(">", level) & " " & msg
```
characters are easier to count visually than spaces, and this would reuse the same rendering as used in `--processing` (refs https://github.com/nim-lang/Nim/pull/18372) for a similar purpose (rendering a tree of locations)

quantimnot replied:
What about using box symbols?
```
└─ lineinfo Error: reason
   ├─ lineinfo Hint: reason
   │  └─ lineinfo Hint: reason
   └─ lineinfo Hint: reason
```

timotheecour replied:
that'd be fine by me but seems harder to implement? it's ascii right? could possibly be implemented in a followup PR?

quantimnot replied:
Not 7-bit US-ASCII. But could be:
```
`-- lineinfo Error: reason
    |-- lineinfo Hint: reason
    |   `-- lineinfo Hint: reason
    `-- lineinfo Hint: reason
```

I'll try to get to it in a later PR (maybe)

timotheecour replied:
ya let's defer to another PR, because my immediate reaction would be: "please generalize in an API that can be reused", even if in some private-for-now module

quantimnot replied:
Only whitespace before a diagnostic message is supported by the VSCode Nim plugin.

Relevant code:
https://github.com/pragmagic/vscode-nim/blob/0272a0544ddf147fde98f8a8f2b624b7dcb19eb3/src/nimBuild.ts#L86

I can submit a PR for it, but I'd first like to know for sure that leading '>' will be the standard for structured messages. This affects all tooling that parses the messages. I think I read somewhere that tooling support is a priority this year or something.

timotheecour replied:
i think we can move forward with this PR, and then in followup work either make vscode nim plugin be aware of `>`, or change back to ' ' here as last resort, but the rationale is that counting visibe chars (eg >) is easier.

There will likely be other static analysis messages done in the compiler that can be written this way so it's good to have 1 way to represent those.

I also wonder whether vscode plugin correctly handles things like the cyclic import error message, or other various multiline messages.

quantimnot replied:
I'm fine to move ahead.

The pragmagic/vscode-nim plugin doesn't seem to support much other than single line diagnostics. I'm not sure of the saem/vscode-nim plugin (just learned of it).

The mutation side effect message rendered by `varpartitions.$` also won't be parsed. That should maybe be combined with this side effect diagnostic message in  a later PR.

timotheecour commented on tests/effects/tdiagnostic_messages.nim:
```diff
@@ -0,0 +1,32 @@
+        tdiagnostic_messages.nim(22, 6) Hint: 'indirectCallViaPointer' called by a `noSideEffect` routine, but has side effects
+          tdiagnostic_messages.nim(23, 32) Hint: 'indirectCallViaPointer' calls a routine with side effects via pointer indirection
+      tdiagnostic_messages.nim(28, 8) Hint: 'callWithSideEffects' calls 'echo' which has side effects
+        ../../lib/system.nim(2004, 6) Hint: 'echo' called by a `noSideEffect` routine, but has side effects
```
CI fails on windows because of this:
```
2021-07-04T17:29:51.2842395Z -        ../../lib/system.nim(2004, 6) Hint: 'echo' called by a `noSideEffect` routine, but has side effects
2021-07-04T17:29:51.2844010Z +        ..\..\lib\system.nim(2004, 6) Hint: 'echo' called by a `noSideEffect` routine, but has side effects
```


this could be fixed by `--filenames:canonical`, but then you'd have a 2nd problem: the test would become flaky as changes in system would break this test.

instead simply add a local proc in this file:
```nim
proc myEcho(a: string) {.sideEffect.} = discard
```

timotheecour commented on tests/effects/tdiagnostic_messages.nim:
```diff
@@ -0,0 +1,32 @@
+  nimout: '''  tdiagnostic_messages.nim(32, 30) Hint: 'a' calls 'callWithSideEffects' which has side effects
+    tdiagnostic_messages.nim(24, 6) Hint: 'callWithSideEffects' called by a `noSideEffect` routine, but has side effects
+      tdiagnostic_messages.nim(26, 34) Hint: 'callWithSideEffects' calls 'indirectCallViaVarParam' which has side effects
+        tdiagnostic_messages.nim(20, 6) Hint: 'indirectCallViaVarParam' called by a `noSideEffect` routine, but has side effects
```
```suggestion
        tdiagnostic_messages.nim(20, 6) Hint: 'indirectCallViaVarParam' called by a `.sideEffect` routine
```
simpler = better.

ditto with other messages, make use of `.sideEffect` to simplify the wording

quantimnot replied:
```Hint: 'indirectCallViaVarParam' called by a `.sideEffect` routine``` is confusing to me.
What does `.sideEffect` mean?

timotheecour replied:
`sideEffect` is a valid pragma, opposite of noSideEffect:
```nim
proc echo*(x: varargs[typed, `$`]) {.magic: "Echo", tags: [WriteIOEffect], benign, sideEffect.}
```

and `.foo` is a shortcut for a `{.foo.}` pragma that is used in a growing number of places, it's a convention. ask araq :)

quantimnot replied:
I wanted to communicate that 'indirectCallViaVarParam'  is trying to be used by a noSideEffect routine. I guess I'm implying a fix of changing 'indirectCallViaVarParam' to be `.noSideEffect` or that the `.noSideEffect` caller needs to be changed to `.sideEffect`.

How would you communicate this to  a new Nim programmer?

timotheecour replied:
it should be obvious if you read the whole error message, which has all the context.
The problem with overly verbose error messages is it drowns the essential part (the things that vary from instance to instance, ie the sym name here) into noise (the things that don't vary). So you might help a tiny bit a confused user the the 1st time they read it, but you'll hurt readability for the next N times you'll see this error msg.

quantimnot commented:
The terminal rendering's structure gives context to the hints about what `noSideEffect` routine is trying to use routines with side effects.
The VSCode hints don't provide that context. Should I change them to someting like:
```Hint: 'indirectCallViaVarParam' called by `a` ```

timotheecour commented on tests/effects/tdiagnostic_messages.nim:
```diff
@@ -0,0 +1,32 @@
+  file: "tdiagnostic_messages.nim"
+  line: 31
+"""
+
```
> The terminal rendering's structure gives context to the hints about what noSideEffect routine is trying to use routines with side effects.
The VSCode hints don't provide that context. Should I change them to someting like:
Hint: 'indirectCallViaVarParam' called by `a`

(i prefer threading so i moved your question here)

i don't use VSCode, can you explain further with more context? change to
```
Hint: 'indirectCallViaVarParam' called by `a`
```
from what?

quantimnot replied:
The VSCode screenshot above shows the message ```'globalVar' accessed by a `noSideEffect` routine```.
It seems difficult for a user to answer which `.noSideEffect` routine is trying to use it.
That information is available in the terminal because of the tree structure.

timotheecour replied:
got it.
```
Hint: 'callWithSideEffects' called by a `noSideEffect` routine, but has side effects
```
=>
```
Hint: 'callWithSideEffects' called by `.noSideEffect` `a`
```

i think is crips and has all the information (adding 
```
Hint: `.sideEffect` 'callWithSideEffects' called by `.noSideEffect` `a`
```

would be possible but kind of redundant IMO, at this point, up to you)

quantimnot committed:
Applied PR review feedback:
  - moved collection of side effects from TSym to TContext
  - used pragma shorthand form `.sideEffect` and `.noSideEffect` in messages
  - added leading '>' to structured messages for readability
  - changed `sempass2.markSideEffect` to a proc
  - replaced `system.echo` in the test to make the test compatible with Windows

timotheecour commented on tests/effects/tdiagnostic_messages.nim:
```diff
@@ -1,22 +1,26 @@
+  nimoutFull: true
+  action: "reject"
+  cmd: "nim r --hint:Conf:off $file"
+  nimout: '''tdiagnostic_messages.nim(35, 6) Error: 'a' can have side effects
```
optional nit: https://nim-lang.github.io/Nim/nep1.html
> Prefer multiline triple quote literals to start with a newline; it's semantically identical (it's a feature of triple quote literals) but clearer because it aligns with the next line:
[...]

quantimnot replied:
Nit pick all you want my friend!

timotheecour reviewed (approved):
LGTM, excellent addition

RSDuck commented:
please rather look at https://github.com/saem/vscode-nim as it's [error parsing code]( https://github.com/saem/vscode-nim/blob/main/src/nimvscode/nimBuild.nim#L125) has gotten some fixes (and also feature additions). https://github.com/pragmagic/vscode-nim is somewhat dead.

quantimnot committed:
Applied NEP1 formatting suggestion

quantimnot commented:
Don't merge. I just notice a small bug. Give me a minute.

quantimnot committed:
fixed? message

quantimnot commented:
@timotheecour
Please take another look at the message changes I just made. I'm exhausted and my mind isn't working.
Ready to merge if it's ok.

@RSDuck
I just submitted a PR to support the leading '>' in the diagnostic messages.

timotheecour commented on tests/effects/tdiagnostic_messages.nim:
```diff
@@ -2,19 +2,20 @@ discard """
+>>>> tdiagnostic_messages.nim(27, 6) Hint: 'indirectCallViaPointer' called by `.noSideEffect` 'a'
+>>>>> tdiagnostic_messages.nim(28, 32) Hint: 'indirectCallViaPointer' calls routine via pointer indirection
+>>> tdiagnostic_messages.nim(33, 10) Hint: 'callWithSideEffects' calls `.sideEffect` 'myEcho'
+>>>> tdiagnostic_messages.nim(24, 6) Hint: 'myEcho' called by `.noSideEffect` 'a'
```
I much prefer the previous messages before last commit; the new way is confusing as you're omitting the direct links; in more complex cases (eg with templates involved or deeper stacks), it'll be hard to tell how is X called by Y because it wouldn't show intermediate links.

>  just notice a small bug

which one?

quantimnot replied:
Yeah :)
It had said ```Hint: 'globalVar' accessed by `.noSideEffect` 'callWithSideEffects'```.
`callWithSideEffects` is `.sideEffect`

I just dropped the `.noSideEffect` if that's fine.

timotheecour replied:
good now :)

Araq commented on compiler/semdata.nim:
```diff
@@ -159,6 +159,7 @@ type
     exportIndirections*: HashSet[(int, int)] # (module.id, symbol.id)
     importModuleMap*: Table[int, int] # (module.id, module.id)
     lastTLineInfo*: TLineInfo
+    sideEffects*: Table[int, seq[(TLineInfo, PSym)]] # symbol.id index
```
Looks much too expensive for any realistically sized project. What is the memory consumption like for bootstrapping?

quantimnot replied:
```./koch boot``` (take third result):
Before:
```192.331s; 839.465MiB```
After: 
```192.473s; 839.906MiB```

```time -l nim c -d=release -o=bin/nim compiler/nim``` (run twice; take second result):
Before:
```709918720  maximum resident set size```
After: 
```711106560  maximum resident set size```

timotheecour replied:
* don't measure with `./koch boot`, at least measure with `./koch boot -d:release`, but ideally should be more granular, by computing memory+time for `nim c --compileOnly compiler/nim` for a nim build before this PR vs built after this PR (both built with -d:release)
* also informative would be instrumenting code to compute the sum of `sideEffects.len` for all PContext objects created

I would guess the cost is not significant, but it should be indeed verified.

timotheecour replied:
* compile times are indistinguishable, measuring best of 4 for `nim c --compileOnly compiler/nim` before vs after PR using a nim compiler compiled with -d:release 
* memory usage difference is minimal in the reported stats, but here's a more precise analysis via this small instrumentation:
https://gist.github.com/timotheecour/5b8b1f1602ce3876ba38b189fe14e6a6
(totNode: 13231319, totSym: 321575, sideEffectsMax: 687, sideEffectsTot: 20367)
nim --eval:'echo 13231319/20367'
649.6449648941916

furthermore, the Table uses contiguous memory, unlike the memory allocated for PSym, PNode

=> conclusion: it's not costly.
PNode allocations almost always dominates costs, which is why it'd be worthwhile to move the sparsely used comment node to a separate Table as I've argued elsewhere.

quantimnot committed:
Revert "fixed? message"

This reverts commit ff203422302bb66312172cfdddb6211a35422380.

quantimnot committed:
simpler message

quantimnot commented:
Current rendering with saem/vscode-nim:
![msg](https://user-images.githubusercontent.com/54247259/124602168-94bdd080-de36-11eb-971b-52b9097a7d2c.png)

quantimnot commented:
I don't know why the checks failed:
```
cmd: git -C csources_v1 checkout a8a5241f9475099c823cfe1a5e0ca4022ac201ff
fatal: reference is not a tree: a8a5241f9475099c823cfe1a5e0ca4022ac201ff
```

quantimnot committed:
formatting long proc sig for NEP1

Araq commented:
> I don't know why the checks failed

Me neither, I'm investigating...

timotheecour commented:
> I don't know why the checks failed

retry now that https://github.com/nim-lang/Nim/pull/18427 was merged

quantimnot commented:
Don't merge. I getting an exception while using nimsuggest:
```
/Volumes/origin/ws/ports/rdn/org.nim-lang.nim/origin.src/compiler/sempass2.nim(292) listSideEffects
/Volumes/origin/ws/ports/rdn/org.nim-lang.nim/origin.src/lib/pure/strutils.nim(2698) addf
/Volumes/origin/ws/ports/rdn/org.nim-lang.nim/origin.src/lib/pure/strutils.nim(2638) invalidFormatString
Error: unhandled exception: invalid format string [ValueError]
```

quantimnot committed:
fixed double string formatting bug

timotheecour commented on compiler/sempass2.nim:
```diff
@@ -289,7 +289,7 @@ proc listSideEffects(result: var string; s: PSym; cycleCheck: var IntSet;
 proc listSideEffects(result: var string; s: PSym; conf: ConfigRef; context: PContext) =
   var cycleCheck = initIntSet()
-  result.addf "'$#' can have side effects\n" % s.name.s
+  result.addf("'$#' can have side effects\n", s.name.s)
```
nice catch; we should probably issue a warning error when addf is used with varargs.len == 0

quantimnot replied:
Good idea. I'll submit a PR for it. Warning or error?

timotheecour replied:
warnings are more flexible because they can be turned into errors, but not vice versa, and it's easy to change whether it defaults as a warningAsError:warnSuspiciousAddf:on|off so it gives a natural migration path without breaking change

I'm curious why would nimsuggest crash but not nim though, any clues? or, what was the command? wasn't code before your fix at least valid ?

quantimnot replied:
A proc ident `$` was formatted into the first string and then that was passed to `strutils.addf`.

quantimnot replied:
It was invalid code because of side effects that encouraged me to make this PR.

timotheecour replied:
> Good idea. I'll submit a PR for it. Warning or error?

see also https://github.com/nim-lang/Nim/pull/14282 which may touch related code.
Not that it'd block your PR, just good to be aware of it.

quantimnot replied:
I don't know what I was thinking about putting a warning in there for this. It's not the compiler.
Is there a compile time method of checking a varargs' length? Should there be?
```
proc a(b: varargs[string]) =
  when b.len == 0: {.warning: "I need args!".}
  discard
```

quantimnot replied:
I just found `system.varargsLen` but it seems to be broken.
```
proc a(b: varargs[string]) =
  when varargsLen(b) > 0:
    echo(b)
  else:
    echo("nope")
a()
```
Output: ```[]```

Also:
```
proc a(b: varargs[string]) =
  when varargsLen(kasdjhflkasjdhflkasdjflaisudfyaisduflasdfjhalsjdhfaljskdfh) > 0:
    echo(b)
  else:
    echo("nope")
a()
```
Output: ```[]```

timotheecour replied:
`varargsLen` works inside templates, not procs, refs https://github.com/nim-lang/Nim/issues/18463

let's discuss this in https://github.com/timotheecour/Nim/issues/780 instead to avoid discussing something unrelated to this PR

timotheecour commented:
@araq is there anything else blocking this PR to be merged? there has been no pending review comment since 8 days ago

Leaving PRs open for a long time guarantees bitrot and eventual demotivation from PR author, eventually joining the train of the 200+ open abandoned PRs where original author gave up, despite being in a mergeable state at some point.

the concern regarding "how costly is this" was answered in https://github.com/nim-lang/Nim/pull/18418#discussion_r664799321

Araq commented:
Sorry, it said

> Don't merge. I getting an exception while using nimsuggest:

Will test it on my machine wrt memory consumption and then merge it.

timotheecour commented:
> Sorry, it said [...]

that was fixed in the "fixed double string formatting bug" commit;
if anything, this proves that the "PTAL" label is a good idea (author can add it when ready for next round/merge, anyone can remove it when comments are added / some action is needed on the PR author part)

Araq commented:
Can confirm that the memory and CPU consumption is acceptable.

Araq commented:
Btw this needs a changelog entry.
